### ISSUE TITLE
CSS order breaks with the combination of server and client components

### ISSUE THREAD

**Issue by Netail** [vercel/next.js#65480]
### Link to the code that reproduces this issue

https://github.com/Netail/app-dir-css-order

### To Reproduce

1. `yarn dev` & open 'localhost:3000/'
2. See 1 yellow and 1 blue component
3. Uncomment the 'use client' in the navigation-wrapper.tsx
4. Instead of 1 yellow & 1 blue component it becomes 2 blue ones (Even after a reload & restart)

### Current vs. Expected behavior

CSS streaming of client components mess with the order of the css bundle, resulting in being appended later and thus overriding certain server side styling

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 10
Binaries:
  Node: 20.8.1
  npm: 10.5.0
  Yarn: 1.22.22
  pnpm: N/A
Relevant Packages:
  next: 14.3.0-canary.24 // There is a newer canary version (14.3.0-canary.45) available, please upgrade!
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Webpack

### Which stage(s) are affected? (Select all that apply)

next dev (local)

**Comment by Netail** [comment]
cc: @sokra as you've worked a lot on the CSS bundling part

https://github.com/vercel/next.js/pull/63157 fixed a big issue we had before, but it seems like there's still an issue left regarding the css order when using both client & server components :(

**Comment by iamlinkus** [comment]
I think this is also related and is a critical bug: https://github.com/vercel/next.js/issues/64773#issuecomment-2091386017

**Comment by Netail** [comment]
> I think this is also related and is a critical bug: https://github.com/vercel/next.js/issues/64773

Agreed on this being a critical bug, because this has huge impact on css rendering

**Comment by Netail** [comment]
A bit more info found here; https://github.com/vercel/next.js/issues/51030#issuecomment-2220507886

**Comment by peerjollux** [comment]
Is there any more info about this issue? I'm currently on 14.2.0.canary.28 but I'm still having issue CSS ordering issues when using Client components.

**Comment by samcx** [comment]
@Netail Can confirm this is still broken with `next dev` (with `--turbopack` or not)

**Comment by Netail** [comment]
Might know a fix, using the webpack bundle layer in the css modules hash. This way the classnames won't conflict between server & client components
